client/src/main/java/tds/assessment/Segment.java

```
@@ -138,70 +157,112 @@ public Instant getFieldTestStartDate() {
-    public void setSelectionAlgorithm(Algorithm selectionAlgorithm) {
-        this.selectionAlgorithm = selectionAlgorithm;
-    }
+    /**
+     * @return All the {@link tds.assessment.Form}s associated with this segment.
+     * <p>
+     *     This getter/setter is only used by {@link tds.assessment.Segment}s that are configured for the "fixedform"
+     *     algorithm.  Segments configured for the "adaptive2" algorithm do not have forms, meaning this can return
+     *     null.
+     * </p>
+     */
+    public List<Form> getForms() {
 
-    public void setStartAbility(float startAbility) {
-        this.startAbility = startAbility;
-    }
+        if (forms == null) {
+            forms = new ArrayList<>();
+        }
 
-    public void setSubject(String subject) {
-        this.subject = subject;
-    }
+        return forms;
 
-    public void setAssessmentKey(String assessmentKey) {
-        this.assessmentKey = assessmentKey;
-    }
-
-    public void setPosition(int position) {
-        this.position = position;
     }
 
-    public void setMinItems(int minItems) {
-        this.minItems = minItems;
+    public void setForms(List<Form> forms) {
+        this.forms = forms;
     }
 
-    public void setMaxItems(int maxItems) {
-        this.maxItems = maxItems;
-    }
+    /**
+     * Choose the {@link tds.assessment.Form} for the specified language.
+     * <p>
+     *     This method is only used by {@link tds.assessment.Segment}s that are configured for the "fixedform"
+     *     algorithm.  Segments configured for the "adaptive2" algorithm do not have forms, meaning this can return
+     *     null.
+     * </p>
+     *
+     * @param languageCode The student's specified language
+     * @return The {@link tds.assessment.Form} for the specified language
+     * @throws java.lang.IllegalArgumentException if the languageCode does not match any of the forms contained within
+     * this Segment
+     */
+    public Form getForm(String languageCode) {
```

I think this should return an Optional and not throw.  How is one to know which languages are acceptable?  That or this class should provide a `boolean isLanguageSupported()` method to allow a caller to check before trying to get the form for an unsupported language.

---

@millerts The languageCodes are part of the parent Assessment, presumably one of those codes will be used here.

---

I understand.  I just don't think we need to throw an Exception and instead can return an Optional.empty() and allow the caller to handle it.

---

Unfortunately I don't believe it can return Optional as we are restricted to Java 7 for code in the client jars. But I agree with returning empty lists or null and letting the callers handle it.

---

Unfortunately since this class is in the `client` library, returning an `Optional` will be a bit of a pain.  We could:

* Add guava (or some equivalent) library to get an `Optional` and some functional stuff (that may or may not be useful)
* Return `null`
* Return a new, empty `Form`

Also, isn't it exceptional to ask for a form with a language code the `Segment` doesn't know about?

---

Pursuant to our conversation that @burzum619 is going to change this method to handle the updated information we received about form selection from SBAC, I'm going to leave this the way it is for now.